syntax = "proto3";

package client;

// Messages for peer2peer handshake
message JoinRequest {
  string address = 1;
}
message JoinResponse {
  repeated string peer_list = 1; // list of known peer addresses
}

// Messages for ping
message PingRequest {
  double t0 = 1;  // client send time
}
message PingResponse {
  double t1 = 1;  // server receive time
  double t2 = 2;  // server send time
}

// Messages for music control
message MusicRequest {
  string action = 1;       // "play", "pause", "resume", "stop"
  float wall_clock = 2;    // wall clock time of action
  float wait_time = 3;     // wait time in milliseconds
  int32 position = 4;      // position of song
}

message MusicResponse {}

// Messages for querying position
message GetPositionRequest {}
message GetPositionResponse {
  int32 position = 1;
}

// Define the service
service ClientHandler {
  rpc Join(JoinRequest) returns (JoinResponse);
  rpc Ping(PingRequest) returns (PingResponse);
  rpc SendMusicCommand(MusicRequest) returns (MusicResponse);
  rpc GetPosition(GetPositionRequest) returns (GetPositionResponse);
}